onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h

```diff
@@ -0,0 +1,706 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+#include "tree_ensemble_aggregator.h"
+
+namespace onnxruntime {
+namespace ml {
+
+template <typename ITYPE, typename OTYPE>
+class TreeEnsembleCommon {
+ public:
+  int64_t n_targets_or_classes_;
+
+ protected:
+  std::vector<OTYPE> base_values_;
+  POST_EVAL_TRANSFORM post_transform_;
+  AGGREGATE_FUNCTION aggregate_function_;
+  int64_t n_nodes_;
+  TreeNodeElement<OTYPE>* nodes_;
+  std::vector<TreeNodeElement<OTYPE>*> roots_;
+
+  int64_t max_tree_depth_;
+  int64_t n_trees_;
+  bool same_mode_;
+  bool has_missing_tracks_;
+  int omp_tree_;
+  int omp_N_;
+
+ public:
+  TreeEnsembleCommon(int omp_tree, int omp_N);
+  ~TreeEnsembleCommon();
+
+  void init(
+      const std::string& aggregate_function,
+      std::vector<OTYPE> base_values,
+      int64_t n_targets_or_classes,
+      std::vector<int64_t> nodes_falsenodeids,
+      std::vector<int64_t> nodes_featureids,
+      std::vector<OTYPE> nodes_hitrates,
+      std::vector<int64_t> nodes_missing_value_tracks_true,
+      const std::vector<std::string>& nodes_modes,
+      std::vector<int64_t> nodes_nodeids,
+      std::vector<int64_t> nodes_treeids,
+      std::vector<int64_t> nodes_truenodeids,
+      std::vector<OTYPE> nodes_values,
+      const std::string& post_transform,
+      std::vector<int64_t> target_class_ids,
+      std::vector<int64_t> target_class_nodeids,
+      std::vector<int64_t> target_class_treeids,
+      std::vector<OTYPE> target_class_weights);
+
+  std::string runtime_options();
+
+  int omp_get_max_threads();
+
+  void compute(const Tensor* X, Tensor* Z, Tensor* label) const;
+
+ protected:
+  void init_c(
+      const std::string& aggregate_function,
+      const std::vector<OTYPE>& base_values,
+      int64_t n_targets_or_classes,
+      const std::vector<int64_t>& nodes_falsenodeids,
+      const std::vector<int64_t>& nodes_featureids,
+      const std::vector<OTYPE>& nodes_hitrates,
+      const std::vector<int64_t>& nodes_missing_value_tracks_true,
+      const std::vector<std::string>& nodes_modes,
+      const std::vector<int64_t>& nodes_nodeids,
+      const std::vector<int64_t>& nodes_treeids,
+      const std::vector<int64_t>& nodes_truenodeids,
+      const std::vector<OTYPE>& nodes_values,
+      const std::string& post_transform,
+      const std::vector<int64_t>& target_class_ids,
+      const std::vector<int64_t>& target_class_nodeids,
+      const std::vector<int64_t>& target_class_treeids,
+      const std::vector<OTYPE>& target_class_weights);
+
+  TreeNodeElement<OTYPE>* ProcessTreeNodeLeave(
+      TreeNodeElement<OTYPE>* root, const ITYPE* x_data) const;
+
+  template <typename AGG>
+  void compute_agg(const Tensor* X, Tensor* Z, Tensor* label, const AGG& agg) const;
+};
```

Comment:
Please use references for arguments that must be provided (i.e. nullptr is not valid input).
[https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-ptr-ref](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-ptr-ref)